lib/ood_core/batch_connect/templates/vnccontainer.rb

```diff
@@ -0,0 +1,248 @@
+
+                # Check that Xvnc process is running, if not assume it died and
+                # wait some random period of time before restarting
+                kill -0 ${VNC_PID} 2>/dev/null || sleep 0.$(random_number 1 9)s
+
+                # If running, then all is well and break out of loop
+                kill -0 ${VNC_PID} 2>/dev/null && break
+              done
+
+              # If we fail to start it after so many tries, then just give up
+              kill -0 ${VNC_PID} 2>/dev/null || clean_up 1
+
+              # Parse output for ports used
+              display=$(echo "${VNC_OUT}" | awk -F':' '/^Desktop/{print $NF}')
+              port=$((5900+display))
+
+              echo "Successfully started VNC server on ${host}:${port}..."
+
+              #{super}
+            EOT
+          end
+
+          # Run the script under the VNC server's display
+          def run_script
+            %(DISPLAY=:${display} #{super})
+          end
+
+          # After startup the main script, scan the VNC server log file for
+          # successful connections so that the password can be reset
+          def after_script
+            websockify_cmd = context.fetch(:websockify_cmd, "${WEBSOCKIFY_CMD:-/opt/websockify/run}").to_s
+
+            <<-EOT.gsub(/^ {14}/, "")
+              #{super}
+
+              # Launch websockify websocket server
+              module load #{container_module}
+              echo "Starting websocket server..."
+              websocket=$(find_port)
+              #{container_command} exec instance://#{instance_name} #{websockify_cmd} -D ${websocket} localhost:${port}
+
+              # Set up background process that scans the log file for successful
+              # connections by users, and change the password after every
+              # connection
+              echo "Scanning VNC log file for user authentications..."
+              while read -r line; do
+                if [[ ${line} =~ "Full-control authentication enabled for" ]]; then
+                  change_passwd
+                  create_yml
+                fi
+              done < <(tail -f --pid=${SCRIPT_PID} "#{vnc_log}") &
+            EOT
+          end
+
+          # Clean up the running VNC server and any other stale VNC servers
+          def clean_script
+            <<-EOT.gsub(/^ {14}/, "")
+              #{super}
+              module load #{container_module}
+
+              #{vnc_clean}
+              [[ -n ${display} ]] && vncserver -kill :${display}
+              #{container_command} instance stop #{instance_name}
+            EOT
+          end
+
+          # Log file for VNC server
+          def vnc_log
+            context.fetch(:vnc_log, "vnc.log").to_s
+          end
+
+          # Password file for VNC server
+          def vnc_passwd
+            context.fetch(:vnc_passwd, "vnc.passwd").to_s
+          end
+
+          def instance_name 
+            context.fetch(:instance_name, "desktop").to_s
+          end
```

Comment:
I have another comment above about this being an option.

Aside from that - we should put add some randomization. 

If we remove it as an option entirely it should probably reflect that it's from Open OnDemand so maybe `ood_desktop` or similar as a default?

Reply:
> vnccontainer sets the environmental variable INSTANCE_NAME for use in the app scripts.

I'm now reading this comment.  I'm not as familiar with singularity as other container runtimes.  Would they need this environment variable for the `hostname`?

Reply:
If you want to run things out of the container (in our case, a desktop and ANSYS) then you need the instance name to run `singularity exec instance://<instance name> <command>` in the app scripts.

Reply:
That makes more sense. It'd be nice to embed the job id, but that may be asking too much given how many environment variables we'd have to track down.

Reply:
@johrstrom I'm replacing the user-provided option with a uuid, sound reasonable?